indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java

```diff
@@ -0,0 +1,1021 @@
+              break;
+            default:
+              throw new IAE("Unknown FS Scheme: [%s]", outFSScheme);
+          }
+          HadoopDruidConverterConfig.jsonMapper.writeValue(descriptorStream, finalSegment);
+        }
+      }
+      finally {
+        if (tmpZipFile.exists() && !tmpZipFile.delete()) {
+          log.warn("Could not delete temporary file [%s]", tmpZipFile);
+        }
+      }
+
+      context.progress();
+      context.setStatus("Finished PUSH");
+      final String finalSegmentString = HadoopDruidConverterConfig.jsonMapper.writeValueAsString(finalSegment);
+      context.getConfiguration().set(ConvertingOutputFormat.PUBLISHED_SEGMENT_KEY, finalSegmentString);
+      context.write(new Text("dataSegment"), new Text(finalSegmentString));
+
+      for (File file : files) {
+        context.getCounter(COUNTER_GROUP, COUNTER_WRITTEN).increment(file.length());
+      }
+      context.progress();
+      context.setStatus("Ready To Commit");
+    }
+
+    // We must be consistent with `io.druid.indexer.HadoopDruidIndexerConfig.makeSegmentOutputPath()`
+    private static Path getOutputBasePathForSegment(
+        DataSegment segment,
+        Path baseOutputPath,
+        Configuration configuration
+    )
+        throws IOException
+    {
+      Path path = new Path(baseOutputPath, segment.getDataSource());
+      if (path.getFileSystem(configuration) instanceof DistributedFileSystem) {
+        path = new Path(
+            path,
+            String.format(
+                "%s_%s",
+                segment.getInterval().getStart().toString(ISODateTimeFormat.basicDateTime()),
+                segment.getInterval().getEnd().toString(ISODateTimeFormat.basicDateTime())
+            )
+        );
+        path = new Path(path, segment.getVersion().replace(':', '_'));
+      } else {
+        path = new Path(
+            path,
+            String.format(
+                "./%s_%s",
+                // Since time contains a `:` we specify this is a relative path or else the URI parser freaks out
+                segment.getInterval().getStart().toString(),
+                segment.getInterval().getEnd().toString()
+            )
+        );
+        path = new Path(path, "./" + segment.getVersion());
+      }
+      return new Path(path, Integer.toString(segment.getShardSpec().getPartitionNum()));
+    }
+
+    @Override
+    protected void setup(Context context) throws IOException, InterruptedException
+    {
+      final File tmpFile = Files.createTempDir();
+      context.getConfiguration().set(TMP_FILE_LOC_KEY, tmpFile.getAbsolutePath());
+    }
+
+    private static URI getURIFromSegment(DataSegment dataSegment)
+    {
+      // There is no good way around this...
+      // TODO: add getURI() to URIDataPuller
+      final Map<String, Object> loadSpec = dataSegment.getLoadSpec();
+      final String type = loadSpec.get("type").toString();
+      final URI segmentLocURI;
+      if ("s3_zip".equals(type)) {
```

Comment:
Currently that capacity resides in the DataPusher. 

I would like to see it reside in the DataPuller as well, because there is already a standard way to get a DataPuller from a LoadSpec. See the comment about URIDataPuller immediately prior.